src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,160 @@
+    private Node<T> first;
+    private Node<T> last;
+    private Node<T> currentNode;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            first = new Node<>(null, value, null);
+            last = first;
+            size++;
+            return;
+        }
+        last = new Node<>(last, value, null);
+        last.prev.next = last;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index is not valid. Index is "
+                    + index + ", but it must be in bounds from 0 to " + size);
+        }
+        if (index > 0 && index < size) {
+            getItem(index);
+            currentNode = new Node<>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = currentNode;
+            currentNode.next.prev = currentNode;
+            size++;
+            return;
+        }
+        if (index == 0 && size > 0) {
+            first = new Node<>(null, value, first);
+            first.next.prev = first;
+            size++;
+        }
+        if (index == size) {
+            add(value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            last = new Node<>(last, list.get(i), null);
+            last.prev.next = last;
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        getItem(index);
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        getItem(index);
+        T setValue = currentNode.value;
+        currentNode.value = value;
+        return setValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        getItem(index);
+        T removedValue = currentNode.value;
+        unLink(currentNode);
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        for (currentNode = first; currentNode != null; currentNode = currentNode.next) {
+            if (currentNode.value == object
+                    || currentNode.value != null && currentNode.value.equals(object)) {
+                unLink(currentNode);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Index is not valid. Index is "
+                    + index + ", but it must be in bounds from 0 to " + size);
+        }
+    }
+
+    private void getItem(int index) {
```

Comment:
```suggestion
    private Node<T> getNodeByIndex(int index) {
```
method name suggests that it will return some value